up-transport-vsomeip/src/transport.rs

```diff
@@ -0,0 +1,566 @@
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Apache License Version 2.0 which is available at
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ ********************************************************************************/
+
+use crate::determine_message_type::{
+    determine_registration_type, determine_send_type, RegistrationType,
+};
+use crate::listener_registry::{
+    find_app_name, find_available_listener_id, free_listener_id, get_extern_fn,
+    insert_into_listener_id_map, map_listener_id_to_client_id, register_listener_id_with_listener,
+    release_listener_id,
+};
+use crate::transport_inner::TransportCommand;
+use crate::transport_inner::{
+    UP_CLIENT_VSOMEIP_FN_TAG_INITIALIZE_NEW_APP_INTERNAL,
+    UP_CLIENT_VSOMEIP_FN_TAG_REGISTER_LISTENER_INTERNAL, UP_CLIENT_VSOMEIP_FN_TAG_SEND_INTERNAL,
+    UP_CLIENT_VSOMEIP_FN_TAG_UNREGISTER_LISTENER_INTERNAL, UP_CLIENT_VSOMEIP_TAG,
+};
+use crate::vsomeip_config::extract_applications;
+use crate::UPTransportVsomeip;
+use crate::{any_uuri, any_uuri_fixed_authority_id, ApplicationName};
+use async_trait::async_trait;
+use log::{error, trace, warn};
+use std::sync::Arc;
+use std::time::Duration;
+use tokio::sync::oneshot;
+use tokio::time::timeout;
+use up_rust::{
+    ComparableListener, UAttributesValidators, UCode, UListener, UMessage, UMessageType, UStatus,
+    UTransport, UUri,
+};
+use vsomeip_sys::extern_callback_wrappers::MessageHandlerFnPtr;
+
+const UP_CLIENT_VSOMEIP_FN_TAG_REGISTER_LISTENER: &str = "register_listener";
+
+const INTERNAL_FUNCTION_TIMEOUT: u64 = 3;
+
+async fn await_internal_function(
+    function_id: &str,
+    rx: oneshot::Receiver<Result<(), UStatus>>,
+) -> Result<(), UStatus> {
+    match timeout(Duration::from_secs(INTERNAL_FUNCTION_TIMEOUT), rx).await {
+        Ok(Ok(result)) => result,
+        Ok(Err(_)) => Err(UStatus::fail_with_code(
+            UCode::INTERNAL,
+            format!(
+                "Unable to receive status back from internal function: {}",
+                function_id
+            ),
+        )),
+        Err(_) => Err(UStatus::fail_with_code(
+            UCode::DEADLINE_EXCEEDED,
+            format!(
+                "Unable to receive status back from internal function: {} within {} second window.",
+                function_id, INTERNAL_FUNCTION_TIMEOUT
+            ),
+        )),
+    }
+}
+
+async fn send_to_inner_with_status(
+    tx: &tokio::sync::mpsc::Sender<TransportCommand>,
+    transport_command: TransportCommand,
+) -> Result<(), UStatus> {
+    tx.send(transport_command).await.map_err(|e| {
+        UStatus::fail_with_code(
+            UCode::INTERNAL,
+            format!(
+                "Unable to transmit request to internal vsomeip application handler, err: {:?}",
+                e
+            ),
+        )
+    })
+}
+
+#[async_trait]
+impl UTransport for UPTransportVsomeip {
```

Comment:
why isn't the struct declared here as well?

Reply:
Put the declaration of the struct at the crate root. Dunno. Seemed reasonable.

Reply:
IMHO there is no good reason to do that. If you want to make the struct accessible from the crate's root then you can simply reexport it in lib.rs. But having the struct declared there makes the code much harder to read and follow. After all, modules are the means to structure the code in Rust ...